R/utilities.R

```diff
@@ -361,60 +363,90 @@ get_ssh_session = function(host="gphost01.bcgsc.ca"){
 gene_to_region = function(gene_symbol,
                           ensembl_id,
                           genome_build = "grch37",
-                          return_as = "region"){
-
+                          return_as = "region",
+                          sort_regions = TRUE,
+                          na_for_genes_not_found = FALSE){
+  
+  stopifnot('`genome_build` parameter must be "grch37" or "grch38"' = genome_build %in% c("grch37", "grch38"))
+  stopifnot('`return_as` parameter must be "region", "bed" or "df"' = return_as %in% c("region", "bed", "df"))
+  
   #set mart based on selected genome projection
   if(genome_build == "grch37"){
     gene_coordinates = grch37_gene_coordinates
     chr_select = paste0(c(c(1:22),"X","Y"))
-  }else if(genome_build == "hg38"){
+  }else{
     gene_coordinates = hg38_gene_coordinates
     chr_select = paste0("chr", c(c(1:22),"X","Y"))
   }
-
+  
   #filter on gene_symbol/ensembl_id
   if(!missing(gene_symbol) && missing(ensembl_id)){
-    gene_coordinates = dplyr::filter(gene_coordinates, hugo_symbol %in% gene_symbol)
+    gene_coordinates = gene_coordinates[ match(gene_symbol, gene_coordinates$hugo_symbol), ]
```

Comment:
What is the benefit/purpose of these changes?

Reply:
This change is to ensure that the gene order remains the same as in the input `gene_symbol`. Take a look at this example:
```
gene_symbol <- c("MYC", "STAT6", "CCND3")
gene_coordinates = grch37_gene_coordinates

gene_coordinates[ match(gene_symbol, gene_coordinates$hugo_symbol), ]
# # A tibble: 3 × 6
#   ensembl_gene_id chromosome     start       end gene_name hugo_symbol
#   <chr>           <chr>          <dbl>     <dbl> <chr>     <chr>      
# 1 ENSG00000136997 8          128747680 128753674 MYC       MYC        
# 2 ENSG00000166888 12          57489191  57525922 STAT6     STAT6      
# 3 ENSG00000112576 6           41902671  42018095 CCND3     CCND3      

dplyr::filter(gene_coordinates, hugo_symbol %in% gene_symbol)
# # A tibble: 3 × 6
#   ensembl_gene_id chromosome     start       end gene_name hugo_symbol
#   <chr>           <chr>          <dbl>     <dbl> <chr>     <chr>      
# 1 ENSG00000166888 12          57489191  57525922 STAT6     STAT6      
# 2 ENSG00000112576 6           41902671  42018095 CCND3     CCND3      
# 3 ENSG00000136997 8          128747680 128753674 MYC       MYC        
```

Reply:
The next line "assigning the hygo_symbol column value of the gene_symbol" is to deal with samples that don't exist. Then we don't miss the name of missing genes. Example:
```
gene_symbol <- c("MYC", "STAT6", "imaginary_gene", "CCND3")
gene_coordinates = grch37_gene_coordinates
gene_coordinates = gene_coordinates[ match(gene_symbol, gene_coordinates$hugo_symbol), ]
gene_coordinates
# A tibble: 4 × 6
#   ensembl_gene_id chromosome     start       end gene_name hugo_symbol
#   <chr>           <chr>          <dbl>     <dbl> <chr>     <chr>      
# 1 ENSG00000136997 8          128747680 128753674 MYC       MYC        
# 2 ENSG00000166888 12          57489191  57525922 STAT6     STAT6      
# 3 NA              NA                NA        NA NA        NA         
# 4 ENSG00000112576 6           41902671  42018095 CCND3     CCND3      

gene_coordinates$hugo_symbol <- gene_symbol
gene_coordinates
# # A tibble: 4 × 6
#   ensembl_gene_id chromosome     start       end gene_name hugo_symbol   
#   <chr>           <chr>          <dbl>     <dbl> <chr>     <chr>         
# 1 ENSG00000136997 8          128747680 128753674 MYC       MYC           
# 2 ENSG00000166888 12          57489191  57525922 STAT6     STAT6         
# 3 NA              NA                NA        NA NA        imaginary_gene
# 4 ENSG00000112576 6           41902671  42018095 CCND3     CCND3         
```

Reply:
why would we want those back with all NA? I guess this is meant to alert the user that something went wrong?

Reply:
Thanks for the example, it is clear now. I think this might be addressed by piping this to the `dplyr::arrange()` to match the specified gene order, so we're sticking with dplyr syntax here for consistency

Reply:
I think he's using data.table match here to ensure genes not in the data frame get added but I'm not sure this is really a feature we want. Might be easier just to warn the user about any genes that had no region info available.

Reply:
I agree this can be thrown as a warning to the user and does not need to be present in the output as NA.

Reply:
I got this idea of returning NAs (adding the argument `na_for_genes_not_found`) because I wanted the output to have the same length and order as the input gene names. I needed that to write my function to report CNV and SSM combined status. But I change the code and I'm not using this even in my function. Maybe we should remove this `na_for_genes_not_found` parameter any never return NAs.

Reply:
This sounds good, I agree - and also think we need to implement Ryan's suggestion of giving a warning for any gene that would be NA here, so at least the user is aware of this

Reply:
Changes that I made:
- the argument `na_for_genes_not_found` has been removed and output never includes NAs;
- If genes have no region info available, a message listing these genes is printed, e.g.
```
Some input gene(s) have no region info available. They are:
imaginary_gene.
8 region(s) returned for 9 gene(s)
```
- This part of the code was rewritten with `dplyr` syntax for consistency.